ark/spLDA/processing.py

```diff
@@ -184,3 +187,121 @@ def create_difference_matrices(cell_table, features, training=True, inference=Tr
 
     matrix_dict = {"train_diff_mat": train_diff_mat, "inference_diff_mat": inference_diff_mat}
     return matrix_dict
+
+
+def gap_stat(features, k, clust_inertia, num_boots):
+    """Computes the Gap-statistic for a given k-means clustering model as introduced by
+    Tibshirani, Walther and Hastie (2001).
+
+    Args:
+        features (pandas.DataFrame):
+            A DataFrame of featurized cellular neighborhoods.  Specifically, this is one of the
+            outputs of :func:`~ark.spLDA.processing.featurize_cell_table`.
+        k (int):
+            The number of clusters in the k-means model.
+        clust_inertia (float):
+            The calculated inertia from the k-means fit using the featurized data.
+        num_boots (int):
+            The number of bootstrap reference samples to generate.
+
+    Returns:
+        tuple (float, float)
+
+        - Estimated difference between the the expected log within-cluster sum of squares and
+        the observed log within-cluster sum of squares (a.k.a. the Gap-statistic).
+        - A scaled estimate of the standard error of the expected log
+        within-cluster sum of squares.
+
+    """
+    # Calculate the range of each feature column
+    mins, maxs = features.apply(min, axis=0), features.apply(max, axis=0)
+    n, p = features.shape
+    w_kb = []
+    # Create bootstrapped reference data
+    boots = np.zeros(shape=(n * num_boots, p))
+    boot_ind = np.repeat(np.array(range(num_boots)), n)
```

Comment:
This could gobble up memory and cause a slow down if the number of cells is high/dense.  Might be worth checking the speed for ~30k spoofed cells.

Reply:
Was actually going to touch base with you about your cython code and see if we can find a solution that way.  What say you?

Reply:
Depends.   Quick mini essay:

The time complexity looks like it's fixed at `O(n * r)` where `n` is the cell count and `r` is the bootstrap number, which shouldn't be too bad actually.  The issue can comes with this implementation's spatial complexity, which is currently `O(n * r * p)` where `p` is the feature-marker count.  Python can end up taking a lot more time when it's dealing with a lot of memory.  In particular, the `boots` variable for 30k cells, 100 bootstrap_num, and 30 markers is like 90 million entries, so with the default float64's, that's about 0.7 GB's of allocated memory.  That's almost a whole GB of allocated data that Python has to not throw a tantrum over.

So, if the code is slow in the 30k cell case, that's probably the reason, which means the first fix is to try implementing this with lower spatial complexity.  One version is something like this:

```
# this replaces 'bootstrapped reference data' creation
boot_array = np.zeros((n, p))

# looping looks like this
for b in range(num_boots):
    for i in range(p):
        boot_array[:, i] = np.random.uniform(low=mins[i], high=maxs[i], size=n)
    boo_clust = ...
    ...
```

while this is worse time complexity wise ( `O(n * r * p)` vs `O(n * r)` ) this removes `r` from the spatial complexity, which can end up (surprisingly) saving time.

If it's still slow after something like that, then it's cython time.

Reply:
Using 30k cells, 20 features, and 100 bootstrap samples, the original code took about 200 seconds and the one you suggested took about 184 seconds.

Reply:
That doesn't sound too bad then.  Since those would probably be some of our max settings, I think 180-200s is reasonable.  

That being said, if we really really want this to be lightning fast, the speed improvements from a cython-ization depend on where the 'slow' part is.  Using the `cProfile` module, you can see what function calls are taking up the most cumulative time.  If that ends up being the `np.random.uniform` initializations, then cython could speed that up quite a bit.  If it's the `Kmeans` call (that's my suspicion), then idk if there's much to be done on cython's end.

Reply:
Yeah most of the work is happening in the KMeans call.  I guess it's fine the way it is for now.  It's just an exploratory tool so I don't think there's much value in doing more than 100 bootstrap samples anyway.